### ISSUE TITLE
isinstance edge case resets all type-narrowing

### ISSUE THREAD

**Issue by bijij** [microsoft/pyright#2049]
**Describe the bug**
returning in a `not isinstance` block which validates a type as an instance of a non-builtin type resets all type-narrowing in the funtion scope.

**To Reproduce**
See below code:

pyright returns
```
test.py:14:17 - info: Type of "item.name" is "str"
test.py:15:17 - info: Type of "baz" is "int"
test.py:20:17 - info: Type of "item.name" is "str | None"
test.py:21:17 - info: Type of "baz" is "int | str"
```

interestingly if we replace `Item` with a builtin such as `str` this error no longer occurs.

**Expected behavior**
`item.name` wold still be of type `"str"`
and `baz` would still be of type `"int"`

**Screenshots or Code**
```py
from typing import Optional, Union

class Item:
    name: Optional[str]


baz: Union[int, str]

async def foo(item: Item, object: Union[int, str]) -> None:
    assert item.name is not None
    assert isinstance(baz, int)


    reveal_type(item.name)
    reveal_type(baz)

    if not isinstance(object, Item):
        return

    reveal_type(item.name)
    reveal_type(baz)

```

![image](https://user-images.githubusercontent.com/8677174/124264320-aaea3900-db77-11eb-9749-0d6f6ce7ccf3.png)

**VS Code extension or command-line**
pyright 1.1.154 and pylance 2021.6.3

**Additional context**
mypy reveals the expected types.

**Comment by erictraut** [comment]
The type evaluator knows that `not isinstance(object, Item)` will always evaluate to `True`, so the return statement will always be taken. Any code below it is effectively unreachable from the code above this conditional block, so the code flow engine cannot propagate the narrowed types. It falls back on the un-narrowed (original) types in this case. I think this is reasonable behavior in this case.

I assume that this sample is simplified from some more complicated piece of code. In the original code, does the `not isinstance` check always evaluate to `True` as well? If so, what is the purpose of this check, and why is there code after it if the `return` will always be taken at runtime?

**Comment by bijij** [comment]
That makes sense to me, sorry for taking up your time.

You are correct that the actual code is more complicated, the actual code takes two `Optional` arguments, objects representing data obtained via a web API, (the types are not exhaustively validated by this point in the code) checks that only one of them `is not None` then confirms the type is one of the expected types in the `not isinstance` check, The issue is I had added an additional class not derived from either of the two argument types for future-proofing, as it is expected the API would be sending this additional type through at some point.

Trivial to remove it for now.

as a side note:
![image](https://user-images.githubusercontent.com/8677174/124290976-67072c00-db97-11eb-96e3-99049310184c.png)
I figure getting pyright to conclude `c` cannot be `None` would be fairly difficult in this case?

**Comment by erictraut** [comment]
"Fairly difficult" is an understatement. :) It would require tracking types for each symbol that are conditioned on types of other symbols. In the sample above, it would need to reason that "if `a` is None, then `b` is not None" and vice versa. Then it would need to use that knowledge to compute all possible combinations of `a or b` and union those types together.

I don't know of any type checker in any language that does this. Setting aside the complexity of the logic involved, it would be computationally infeasible except for very simple code samples.

The `object` comes from the generic `T`, which is not bound to a specific type, so it is, in effect, bound to `object`.  There are certain operations where the type evaluator needs to convert a type variable into its concrete form to perform the evaluation. The `or` operator is one of those cases. The LHS of the `or` operator is "concretized".

I think this is all working as expected, so I'm going to close this issue.
